sycl/plugins/unified_runtime/ur/adapters/hip/device.cpp

```diff
@@ -0,0 +1,996 @@
+  }
+  case UR_DEVICE_INFO_VENDOR_ID: {
+#if defined(__HIP_PLATFORM_AMD__)
+    uint32_t VendorId = 4098u;
+#elif defined(__HIP_PLATFORM_NVIDIA__)
+    uint32_t VendorId = 4318u;
+#else
+    uint32_t VendorId = 0u;
+#endif
+    return ReturnValue(VendorId);
+  }
+  case UR_DEVICE_INFO_MAX_COMPUTE_UNITS: {
+    int ComputeUnits = 0;
+    detail::ur::assertion(
+        hipDeviceGetAttribute(&ComputeUnits,
+                              hipDeviceAttributeMultiprocessorCount,
+                              hDevice->get()) == hipSuccess);
+    detail::ur::assertion(ComputeUnits >= 0);
+    return ReturnValue(static_cast<uint32_t>(ComputeUnits));
+  }
+  case UR_DEVICE_INFO_MAX_WORK_ITEM_DIMENSIONS: {
+    return ReturnValue(MaxWorkItemDimensions);
+  }
+  case UR_DEVICE_INFO_MAX_WORK_ITEM_SIZES: {
+    struct {
+      size_t sizes[MaxWorkItemDimensions];
+    } return_sizes;
+
+    int MaxX = 0, MaxY = 0, MaxZ = 0;
+    detail::ur::assertion(hipDeviceGetAttribute(&MaxX,
+                                                hipDeviceAttributeMaxBlockDimX,
+                                                hDevice->get()) == hipSuccess);
+    detail::ur::assertion(MaxX >= 0);
+
+    detail::ur::assertion(hipDeviceGetAttribute(&MaxY,
+                                                hipDeviceAttributeMaxBlockDimY,
+                                                hDevice->get()) == hipSuccess);
+    detail::ur::assertion(MaxY >= 0);
+
+    detail::ur::assertion(hipDeviceGetAttribute(&MaxZ,
+                                                hipDeviceAttributeMaxBlockDimZ,
+                                                hDevice->get()) == hipSuccess);
+    detail::ur::assertion(MaxZ >= 0);
+
+    return_sizes.sizes[0] = size_t(MaxX);
+    return_sizes.sizes[1] = size_t(MaxY);
+    return_sizes.sizes[2] = size_t(MaxZ);
+    return ReturnValue(return_sizes);
+  }
+
+  case UR_DEVICE_INFO_MAX_WORK_GROUPS_3D: {
+    struct {
+      size_t sizes[MaxWorkItemDimensions];
+    } return_sizes;
+
+    int MaxX = 0, MaxY = 0, MaxZ = 0;
+    detail::ur::assertion(hipDeviceGetAttribute(&MaxX,
+                                                hipDeviceAttributeMaxGridDimX,
+                                                hDevice->get()) == hipSuccess);
+    detail::ur::assertion(MaxX >= 0);
+
+    detail::ur::assertion(hipDeviceGetAttribute(&MaxY,
+                                                hipDeviceAttributeMaxGridDimY,
+                                                hDevice->get()) == hipSuccess);
+    detail::ur::assertion(MaxY >= 0);
+
+    detail::ur::assertion(hipDeviceGetAttribute(&MaxZ,
+                                                hipDeviceAttributeMaxGridDimZ,
+                                                hDevice->get()) == hipSuccess);
+    detail::ur::assertion(MaxZ >= 0);
+
+    return_sizes.sizes[0] = size_t(MaxX);
+    return_sizes.sizes[1] = size_t(MaxY);
+    return_sizes.sizes[2] = size_t(MaxZ);
+    return ReturnValue(return_sizes);
```

Comment:
It will deal with it as an array, but this struct will be handled internally in UR in this [`urReturnHelper` class](https://github.com/intel/llvm/blob/663042b04b63cb1830ae12b4442e12af4830764e/sycl/plugins/unified_runtime/ur/ur.hpp#L250) which will be casted to an array.

Reply:
Gotcha.